src/sallyport/mod.rs

```diff
@@ -195,32 +197,95 @@ impl<'a> Cursor<'a> {
     pub unsafe fn alloc<T>(
         self,
         count: usize,
-    ) -> core::result::Result<(Cursor<'a>, &'a mut [T]), ()> {
+    ) -> core::result::Result<(Cursor<'a>, &'a mut [MaybeUninit<T>]), ()> {
         let mid = {
-            let (padding, data, _) = self.0.align_to_mut::<T>();
+            let (padding, data, _) = self.0.align_to_mut::<MaybeUninit<T>>();
 
             if data.len() < count {
                 return Err(());
             }
 
-            padding.len() + size_of::<T>() * count
+            padding.len() + size_of::<MaybeUninit<T>>() * count
         };
 
         let (data, next) = self.0.split_at_mut(mid);
 
-        Ok((Cursor(next), data.align_to_mut::<T>().1))
+        Ok((Cursor(next), data.align_to_mut::<MaybeUninit<T>>().1))
     }
 
-    /// Copies data from a value into a slice using self.alloc().
+    /// Copies data from a slice into a slice using self.alloc().
     #[allow(dead_code)]
-    pub fn copy_slice<T: Copy>(
+    pub fn copy_from_slice<T: 'a + Copy>(
         self,
-        value: &[T],
+        src: &[T],
     ) -> core::result::Result<(Cursor<'a>, &'a mut [T]), ()> {
-        let (c, slice) = unsafe { self.alloc(value.len())? };
-        slice.copy_from_slice(value);
+        let (c, slice) = unsafe { self.alloc::<T>(src.len())? };
+        let slice = unsafe { &mut *(slice as *mut [MaybeUninit<T>] as *mut [T]) };
```

Comment:
See https://doc.rust-lang.org/core/mem/union.MaybeUninit.html#initializing-an-array-element-by-element
which uses:
```rust
unsafe { mem::transmute::<_, [Vec<u32>; 1000]>(data) }
```
but clippy suggested using instead:
```rust
unsafe { &mut *(slice as *mut [MaybeUninit<T>] as *mut [T]) };
```

There is no `slice.assume_init()` because `slice` is `&[MaybeUninit<T>]` and not `MaybeUnit<[T]>`

Reply:
completely overhauled the whole thing